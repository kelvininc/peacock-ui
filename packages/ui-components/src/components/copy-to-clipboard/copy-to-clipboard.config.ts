import { ComputePositionConfig, offset } from '@floating-ui/dom';
import { EIconName } from '../icon/icon.types';
import { ECopyToClipboardState } from './copy-to-clipboard.types';
import { ETooltipPosition } from '../../types';

export const ICON_CONFIGS: Record<ECopyToClipboardState, EIconName> = {
	[ECopyToClipboardState.ReadyToCopy]: EIconName.CopyClipboard,
	[ECopyToClipboardState.Copied]: EIconName.Success
};

export const TOOLTIP_CONFIG: Partial<ComputePositionConfig> = {
	placement: ETooltipPosition.BottomEnd,
	middleware: [
		offset({
			mainAxis: 4,
			crossAxis: -8
		})
	]
};

export const STATE_TRANSITION_DURATION_MS = 2000;
export const UNABLE_TO_COPY_ERROR = 'Unable to copy to clipboard';
