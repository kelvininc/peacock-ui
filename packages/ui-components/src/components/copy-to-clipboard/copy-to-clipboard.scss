@import '../../assets/styles/globals';

:host {
	/**
		@prop --container-width: Width of the container
	**/
	--container-width: fit-content;
}

.copy-to-clipboard-container {
	width: var(--container-width);
}

.copy-to-clipboard-content {
	display: flex;
	justify-content: space-between;
	box-sizing: border-box;
	border-radius: $spacing;
	padding: $spacing-2x;

	&:hover {
		cursor: pointer;
		background-color: kv-color('neutral-6');
	}

	.state-icon {
		align-self: self-end;

		&--success {
			kv-icon {
				--icon-color: #{kv-color('success')};
				--icon-background-color: #{kv-color('neutral-0')};
			}
		}
	}
}
