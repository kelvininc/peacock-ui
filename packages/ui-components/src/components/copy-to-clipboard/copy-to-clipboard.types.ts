export interface ICopyToClipboard {
	/** (required) The text to copy to the clipboard when clicking */
	copiableText: string;
	/** (optional) The suffix to show on the `Copy` tooltip before copying */
	tooltipSuffix?: string;
}

export enum ECopyToClipboardState {
	ReadyToCopy = 'ready-to-copy',
	Copied = 'copied'
}
