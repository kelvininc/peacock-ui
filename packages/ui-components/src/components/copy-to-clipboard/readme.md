# _<kv-copy-to-clipboard>_



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-copy-to-clipboard copiableText="https://kelvin.ai" />

<!-- With custom copy tooltip suffix -->
<kv-copy-to-clipboard copiableText="https://kelvin.ai" tooltipSuffix="Link" />

<!-- With content projection -->
<kv-copy-to-clipboard copiableText="https://kelvin.ai" tooltipSuffix="Link">
	<span>Kelvin Website Link</span>
</kv-copy-to-clipboard>

```


### React

```tsx
import React from 'react';

import { KvCopyToClipboard } from '@kelvininc/react-ui-components';

export const KvCopyToClipboardExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvCopyToClipboard copiableText="https://kelvin.ai" />

		{/*-- With custom copy tooltip suffix --*/}
		<KvCopyToClipboard copiableText="https://kelvin.ai" tooltipSuffix="Link" />

		{/*-- With content projection --*/}
		<KvCopyToClipboard copiableText="https://kelvin.ai" tooltipSuffix="Link">
			<span>Kelvin Website Link</span>
		</KvCopyToClipboard>
	</>
);
```



## Properties

| Property        | Attribute        | Description                                                        | Type     | Default     |
| --------------- | ---------------- | ------------------------------------------------------------------ | -------- | ----------- |
| `copiableText`  | `copiable-text`  | (required) The text to copy to the clipboard when clicking         | `string` | `undefined` |
| `tooltipSuffix` | `tooltip-suffix` | (optional) The suffix to show on the `Copy` tooltip before copying | `string` | `undefined` |


## Shadow Parts

| Part        | Description                   |
| ----------- | ----------------------------- |
| `"content"` | The container for the content |


## CSS Custom Properties

| Name                   | Description            |
| ---------------------- | ---------------------- |
| `--container-width`    | Width of the container |
| `--icon-color-default` | Start icon color       |
| `--icon-color-success` | Success icon color     |
| `--icon-start-opacity` | Start icon opacity     |


## Dependencies

### Depends on

- [kv-tooltip](../tooltip)
- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-copy-to-clipboard --> kv-tooltip
  kv-copy-to-clipboard --> kv-icon
  kv-tooltip --> kv-tooltip-text
  style kv-copy-to-clipboard fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


