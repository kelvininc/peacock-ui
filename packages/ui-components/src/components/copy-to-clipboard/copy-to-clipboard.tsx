import { Component, Host, Prop, State, h } from '@stencil/core';
import { ECopyToClipboardState, ICopyToClipboard } from './copy-to-clipboard.types';
import { ICON_CONFIGS, STATE_TRANSITION_DURATION_MS, DEFAULT_TOOLTIP_CONFIG, DEFAULT_TOOLTIP_DELAY, UNABLE_TO_COPY_ERROR } from './copy-to-clipboard.config';
import { copyTextToClipboard } from '../../utils/clipboard.helper';
import { getTooltipText } from './copy-to-clipboard.utils';
import { ComputePositionConfig } from '@floating-ui/dom';

/**
 * @part content - The container for the content
 * @part icon - The copy icon
 */
@Component({
	tag: 'kv-copy-to-clipboard',
	styleUrl: 'copy-to-clipboard.scss',
	shadow: true
})
export class KvCopyToClipboard implements ICopyToClipboard {
	/** @inheritdoc */
	@Prop({ reflect: true }) copiableText: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) tooltipSuffix?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) tooltipDelay?: number = DEFAULT_TOOLTIP_DELAY;
	/** @inheritdoc */
	@Prop({ reflect: true }) tooltipConfig?: Partial<ComputePositionConfig> = DEFAULT_TOOLTIP_CONFIG;

	@State() copyState: ECopyToClipboardState = ECopyToClipboardState.ReadyToCopy;

	private onTextCopy = async (ev: MouseEvent) => {
		ev.stopPropagation();

		if (this.copyState === ECopyToClipboardState.Copied) return;

		try {
			await copyTextToClipboard(this.copiableText);
			this.copyState = ECopyToClipboardState.Copied;
			setTimeout(() => (this.copyState = ECopyToClipboardState.ReadyToCopy), STATE_TRANSITION_DURATION_MS);
		} catch {
			throw new Error(UNABLE_TO_COPY_ERROR);
		}
	};

	private get tooltipText() {
		return getTooltipText(this.copyState, this.tooltipSuffix);
	}

	render() {
		return (
			<Host>
				<kv-tooltip text={this.tooltipText} options={this.tooltipConfig} delay={this.tooltipDelay} onClick={this.onTextCopy}>
					<div part="content" class="copy-to-clipboard-container">
						<slot></slot>
						<div
							class={{
								'state-icon': true,
								'state-icon--success': this.copyState === ECopyToClipboardState.Copied
							}}
						>
							<kv-icon name={ICON_CONFIGS[this.copyState]} part="icon" />
						</div>
					</div>
				</kv-tooltip>
			</Host>
		);
	}
}
