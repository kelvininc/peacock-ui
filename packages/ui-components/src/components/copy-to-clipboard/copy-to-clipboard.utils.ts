import { ECopyToClipboardState } from './copy-to-clipboard.types';

export const getTooltipText = (state: ECopyToClipboardState, suffix?: string) => {
	if (state === ECopyToClipboardState.Copied) return 'Copied';
	return `Copy ${suffix ? suffix : ''}`.trim();
};
