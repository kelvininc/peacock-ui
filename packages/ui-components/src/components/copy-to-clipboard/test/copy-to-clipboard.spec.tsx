import { SpecPage, newSpecPage } from '@stencil/core/testing';
import { KvCopyToClipboard } from '../copy-to-clipboard';

describe('Copy To Clipboard (unit tests)', () => {
	let page: SpecPage;

	describe('when rendering', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvCopyToClipboard],
				html: `
					<kv-copy-to-clipboard copiable-text='https://kelvin.ai' tooltip-suffix='Link'>
						<span>Click me to copy the link</span>
					</kv-copy-to-clipboard>
				`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
