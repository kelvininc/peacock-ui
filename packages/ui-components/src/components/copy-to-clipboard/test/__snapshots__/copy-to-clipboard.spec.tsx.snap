// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copy To Clipboard (unit tests) when rendering with content projection should match the snapshot 1`] = `
<kv-copy-to-clipboard copiable-text="https://kelvin.ai" tooltip-suffix="Link">
  <mock:shadow-root>
    <div class="copy-to-clipboard-container">
      <kv-tooltip text="Copy Link">
        <div class="copy-to-clipboard-content" part="content">
          <slot></slot>
          <div class="state-icon">
            <kv-icon customclass="icon-16" name="kv-copy-clipboard"></kv-icon>
          </div>
        </div>
      </kv-tooltip>
    </div>
  </mock:shadow-root>
  <span>
    Click me to copy the link
  </span>
</kv-copy-to-clipboard>
`;

exports[`Copy To Clipboard (unit tests) when rendering with custom tooltip suffix should match the snapshot 1`] = `
<kv-copy-to-clipboard copiable-text="https://kelvin.ai" tooltip-suffix="Link">
  <mock:shadow-root>
    <div class="copy-to-clipboard-container">
      <kv-tooltip text="Copy Link">
        <div class="copy-to-clipboard-content" part="content">
          <slot></slot>
          <div class="state-icon">
            <kv-icon customclass="icon-16" name="kv-copy-clipboard"></kv-icon>
          </div>
        </div>
      </kv-tooltip>
    </div>
  </mock:shadow-root>
</kv-copy-to-clipboard>
`;

exports[`Copy To Clipboard (unit tests) when rendering with required props should match the snapshot 1`] = `
<kv-copy-to-clipboard copiable-text="https://kelvin.ai">
  <mock:shadow-root>
    <div class="copy-to-clipboard-container">
      <kv-tooltip text="Copy">
        <div class="copy-to-clipboard-content" part="content">
          <slot></slot>
          <div class="state-icon">
            <kv-icon customclass="icon-16" name="kv-copy-clipboard"></kv-icon>
          </div>
        </div>
      </kv-tooltip>
    </div>
  </mock:shadow-root>
</kv-copy-to-clipboard>
`;
